baseapp/baseapp.go

```diff
@@ -526,53 +510,74 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte, tx sdk.Tx) (result sdk
 		ctx = ctx.WithMultiStore(msCache)
 	}
 
-	finalResult := sdk.Result{}
-	var logs []string
-	for i, msg := range msgs {
-		// Match route.
-		msgType := msg.Type()
-		handler := app.router.Route(msgType)
-		if handler == nil {
-			return sdk.ErrUnknownRequest("Unrecognized Msg type: " + msgType).Result()
-		}
+	// No need to run Msg checks when running CheckTx(), since CheckTx() only needs to validate that fees can be paid
+	// Then when DeliverTx() is called, this will run, preventing us from doing the same checks twice
+	if mode != runTxModeCheck {
 
-		result = handler(ctx, msg)
-
-		// Set gas utilized
-		finalResult.GasUsed += ctx.GasMeter().GasConsumed()
-		finalResult.GasWanted += result.GasWanted
+		// Get the Msg.
+		var msgs = tx.GetMsgs()
+		if msgs == nil || len(msgs) == 0 {
+			return sdk.ErrInternal("Tx.GetMsgs() must return at least one message in list").Result()
+		}
 
-		// Append Data and Tags
-		finalResult.Data = append(finalResult.Data, result.Data...)
-		finalResult.Tags = append(finalResult.Tags, result.Tags...)
+		for _, msg := range msgs {
+			// Validate the Msg
+			err := msg.ValidateBasic()
+			if err != nil {
+				err = err.WithDefaultCodespace(sdk.CodespaceRoot)
+				return err.Result()
+			}
+		}
 
-		// Construct usable logs in multi-message transactions. Messages are 1-indexed in logs.
-		logs = append(logs, fmt.Sprintf("Msg %d: %s", i+1, finalResult.Log))
+		finalResult := sdk.Result{}
+		var logs []string
+		for i, msg := range msgs {
+			// Match route.
+			msgType := msg.Type()
+			handler := app.router.Route(msgType)
+			if handler == nil {
+				return sdk.ErrUnknownRequest("Unrecognized Msg type: " + msgType).Result()
+			}
 
-		// Stop execution and return on first failed message.
-		if !result.IsOK() {
-			if len(msgs) == 1 {
+			result = handler(ctx, msg)
+
+			// Set gas utilized
+			finalResult.GasUsed += ctx.GasMeter().GasConsumed()
+			finalResult.GasWanted += result.GasWanted
+
+			// Append Data and Tags
+			finalResult.Data = append(finalResult.Data, result.Data...)
+			finalResult.Tags = append(finalResult.Tags, result.Tags...)
+
+			// Construct usable logs in multi-message transactions. Messages are 1-indexed in logs.
+			logs = append(logs, fmt.Sprintf("Msg %d: %s", i+1, finalResult.Log))
+
+			// Stop execution and return on first failed message.
+			if !result.IsOK() {
+				if len(msgs) == 1 {
+					return result
+				}
+				result.GasUsed = finalResult.GasUsed
+				if i == 0 {
+					result.Log = fmt.Sprintf("Msg 1 failed: %s", result.Log)
+				} else {
+					result.Log = fmt.Sprintf("Msg 1-%d Passed. Msg %d failed: %s", i, i+1, result.Log)
```

Comment:
Why the `1-` ?

Reply:
If a multimessage tx fails, this makes things easier to debug. Easy to tell from reading error log, which message caused the failure.

Do you mean why isn't it 0-indexed? No reason, we can change it to 0 index.